### PR TITLE
virt-launcher: Use interface for arch specific conversions

### PR DESCRIPTION
**Pull request by jschintag**
Replace long switch case statements for arch specific conversions into separate code files by utilizing an interface.
Fix some code warnings in converter.go

### What this PR does
Before this PR:
Big chunks of arch specific code is mixed with general virt-launcher code
After this PR:
arch specific code is mostly in arch specific files. Exceptions is for trivial paths where only single string values are different and the effort and additional complexity of splitting it would not be worth it.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:
- Some arch specific configuration values are still in the main code
The following alternatives were considered:

Links to places where the discussion took place: 
https://github.com/kubevirt/kubevirt/pull/10490#issuecomment-2024935029
### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (alicefr) [inline pkg/virt-launcher/virtwrap/converter/archconverter.go:31]: nit: what about calling this function `NewArchConvertor`?

File: pkg/virt-launcher/virtwrap/converter/converter_test.go
Comment (alicefr) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go:59]: Can you please split this formatting/styling changes in a separate commit?

**Comment by alicefr** [comment]
@jschintag the code overall looks good, thanks!  Please, split the formatting changes in a separate commit. Would you mind also to add some unit tests for the arch converter?

@andreabolognani @zhlhahaha can you please also take a look to this?

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (alicefr) [inline pkg/virt-launcher/virtwrap/converter/archconverter.go:37]: Please add the validation of the arch string.

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (zhlhahaha) [inline pkg/virt-launcher/virtwrap/converter/archconverter.go:29]: Just curious. Why do some functions start with capital letter but one does not?

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go:37]: I would rather not do that for 2 reasons:
1. Adding the error makes the parts of the code where we call the functions more complicated, as i can't just slot a call to ArchConverter in as a replacement. Additionally especially `isUSBNeeded` does not throw an error itself. Though this is a minor reason
2. My main reason for not throwing an error is that we do not do that anywhere else in the code. In all cases the default behaviour is always to assume that the arch is x86_64. I don't see a reason to change it here, as it was also here the case

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go:29]: Because i copied the name of the parent `isUSBNeeded` function and forgot to capitalize it. Thanks for catching that.

A different question would be should the interface be exported from the package, since it is not used outside of it?

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go]: I have renamed the function

File: pkg/virt-launcher/virtwrap/converter/converter_test.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go:59]: I have split the formatting changes into a separate commit.

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (zhlhahaha) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go:29]: > A different question would be should the interface be exported from the package, since it is not used outside of it?

From my point of view, they are supposed to be private function or interface.

**Comment by zhlhahaha** [comment]
Hi @alicefr 
This PR looks good to me. Thanks @jschintag !

**Comment by fossedihelm** [comment]
Just an idea: How about encapsulating the `ArchConverter` in a new struct (with dedicated creator) that will replace the
`Architecture` in the converterContext.

```
type ConverterContext struct {
	Architecture                    Architecture
	...
}

type Architecture struct {
	arch string
	ArchConverter
}
```
 In this way, we have to initialize it only once:
 ```
 	c := &converter.ConverterContext{
		Architecture:          converter.NewArchitecture(runtime.GOARCH),
		...
	}
 ```
 
 Maybe in that case we can also add a receiver to the `is<ARC>` functions:
 ```
 func (a *Architecture) IsARM64() bool {
	return a.arch == "arm64"
}

func (a *Architecture) IsPPC64() bool {
	return a.arch == "ppc64le"
}

func (a *Architecture) IsAMD64() bool {
	return a.arch == "amd64"
}

func (a *Architecture) GetArch() string {
	return a.arch
}
 ```
 WDYT?

File: pkg/virt-launcher/virtwrap/converter/converter.go
Comment (fossedihelm) [inline pkg/virt-launcher/virtwrap/converter/converter.go:1211]: I think this comment should be moved under the specific method of `IsUSBNeeded` for ppc64le struct

**Comment by jschintag** [comment]
@fossedihelm I did think about moving the archConverter under the ConverterContext, but decided against it because we do not have a constructor function for ConverterContext, so we would need to change this in every part where ConverterContext is created.
That does not mean we can't do it, it would just be a lot more effort for little benefit, since creating an archConverter is a trivial operation. We would essentially change a code in according to my quick search 40 places accross 6 files just to have the call to archconverter look a bit more cleaner.

File: pkg/virt-launcher/virtwrap/converter/converter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/converter.go]: Makes sense, i moved it.

**Comment by jschintag** [comment]
> @jschintag the code overall looks good, thanks! Please, split the formatting changes in a separate commit. Would you mind also to add some unit tests for the arch converter?

@alicefr I have made the split into separate commits as suggested. Regarding the unit tests, i'm not sure what i would need to add here, since all of the functionality of arch converter is already tested explicitly by the unit tests. The only thing that would only be tested by coincidence right now would be `NewArchConverter`. I can add a test for that or did you have more in mind?

**Comment by alicefr** [comment]
> > @jschintag the code overall looks good, thanks! Please, split the formatting changes in a separate commit. Would you mind also to add some unit tests for the arch converter?
> 
> @alicefr I have made the split into separate commits as suggested. Regarding the unit tests, i'm not sure what i would need to add here, since all of the functionality of arch converter is already tested explicitly by the unit tests. The only thing that would only be tested by coincidence right now would be `NewArchConverter`. I can add a test for that or did you have more in mind?

You should check that the defaults for each architecture and methods you have added to the interface are the ones you expect. You can overwrite the GOARCH env variable to test the different architectures.

**Comment by fossedihelm** [comment]
> @fossedihelm I did think about moving the archConverter under the ConverterContext, but decided against it because we do not have a constructor function for ConverterContext, so we would need to change this in every part where ConverterContext is created. That does not mean we can't do it, it would just be a lot more effort for little benefit, since creating an archConverter is a trivial operation. We would essentially change a code in according to my quick search 40 places accross 6 files just to have the call to archconverter look a bit more cleaner.

Yes, there are several places where we initialize the convertercontex, but most of the time they are part of the unit tests.
And yes, the most important benefit will be readability, but not only that: we can also reduce code duplication playing with it in `Defaulter` struct (which defines `Is<ARCH>` functions again).
We can keep it for a follow up PR, or doing it now. Up to you.
Thank you!

**Comment by jschintag** [comment]
Yesterday i have tested adding archConverter to ConverterContext. The main problem is that we would need to ensure that it is always non-empty, as otherwise calling any function that depends on it results in panics.
So even though all calls outside of tests are few and always set Architecture, that is not the  case inside tests, and often times ConverterContext is created ad-hoc.
I think the best way to guarantee that archConverter would not be nil would be to outsource creating it to a function, but i think it would be better to leave this to a follow up PR, since it should be looked at more carefully.

Other than that, i pushed the following changes:
- Added unit-test for newArchConverter function
- made all archConverter calls internal to converter package
- Removed SCSIController function, as it is not needed (yet)
- Simplified Power implementation for graphics device as suggested by https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1601217517

Please re-review

**Comment by alicefr** [comment]
There is a typo in the archconverter_test file name

File: pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go
Comment (alicefr) [inline pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go:24]: I would expect here some tests for `addGraphicsDevice` and `isUSBNeeded` too

Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go:24]: Those are already covered for each arch in `converter_test.go` by comparing them to pre-defined expected xml definitions. As such i don't think it makes much sense to duplicate this code again.
After all the code inside those functions did not change, we just now call it a bit differently. As such it is covered by existing tests.

File: pkg/virt-launcher/virtwrap/converter/BUILD.bazel
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/BUILD.bazel]: fixed

File: pkg/virt-launcher/virtwrap/converter/converter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/converter.go]: I have removed the function and called archConverter directly.

File: pkg/virt-launcher/virtwrap/converter/converter.go
Comment (fossedihelm) [inline pkg/virt-launcher/virtwrap/converter/converter.go:95]: just an idea:
Maybe we can move this inside `archconverter.go` file

**Review by fossedihelm — COMMENTED** [review]
Thank you for your patience.
We are almost there, just small minor comments.

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go]: I have dropped it from `isUSBNeeded`. I have left it in `addGraphicsDevice`, since even though only amd64 needs a single boolean from it, giving the function a pointer is just as cheap as copying the boolean.

File: pkg/virt-launcher/virtwrap/converter/converter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/converter.go]: I have moved it. It can be argued that it makes sense in both files, but in the end the archconverter is using it.

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (fossedihelm) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go]: Fair enough, thanks

**Review by fossedihelm — APPROVED** [review]
Thank you for this!
Nothing else from my side.
Good job

**Comment by alicefr** [comment]
/approve
/hold
Put on hold if @iholder101 wants to take a look, otherwise, feel free to unhold
@jschintag thanks for the work!

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (iholder101) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go:41]: I also think that validation would be valuable here.
@jschintag I would suggest something like:
```go
	switch {
	case isARM64(arch):
		return archConverterARM64{}
	case isPPC64(arch):
		return archConverterPPC64{}
	case isAMD64(arch):
		return archConverterAMD64{}
	default:
		log.Warningf("Trying to create an arch converter from an unknown arch: %s. falling back to AMD64", arch)
		return archConverterAMD64{}
```

This way you will able to do things like `ArchConverter(arch).isUSBNeeded()`, and the default would indeed be AMD64. But, on the other hand, an unknown arch being passed would still be somewhat visible.

My main concern is that without this a bug would probably be very hard to solve.

WDYT?

File: pkg/virt-launcher/virtwrap/converter/arm64.go
Comment (iholder101) [inline pkg/virt-launcher/virtwrap/converter/arm64.go:1]: Can you add unit tests for the three converters?

File: pkg/virt-launcher/virtwrap/converter/arm64.go
Comment (iholder101) [inline pkg/virt-launcher/virtwrap/converter/arm64.go:27]: Can you please add a `var _ = archConverter(&archConverterARM64{})`?
This will ensure in compilation time that the struct implements the interface and fail otherwise.
Please do so for all 3 converters.

**Review by iholder101 — COMMENTED** [review]
Thank you @jschintag! Looks great!
See comments below

File: pkg/virt-launcher/virtwrap/converter/arm64.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/arm64.go:1]: See https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1603027526

File: pkg/virt-launcher/virtwrap/converter/arm64.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/arm64.go:27]: The `newArchConverter` function already explicitly returns `archConverter`, so this is already covered.
For example if you comment out isUSBNeeded:
```
pkg/virt-launcher/virtwrap/converter/archconverter.go:40:10: cannot use archConverterARM64{} (value of type archConverterARM64) as archConverter value in return statement: archConverterARM64 does not implement archConverter (missing method isUSBNeeded)
```

**Comment by jschintag** [comment]
@iholder101 Thank you for your review.
Github does not let me quote https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1611334858, but i liked the idea and have implemented it. Though i used the logger from client-go, as it is already widely used in the project.

Would you mind reviewing the changes again?

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (iholder101) [inline pkg/virt-launcher/virtwrap/converter/archconverter.go:46]: typo
```suggestion
		log.Log.Warning("Trying to create an arch converter from an unknown arch: " + arch + ". falling back to AMD64")
```

File: pkg/virt-launcher/virtwrap/converter/arm64.go
Comment (iholder101) [reply pkg/virt-launcher/virtwrap/converter/arm64.go:1]: Thanks for the pointer :+1: Let's discuss it there

File: pkg/virt-launcher/virtwrap/converter/arm64.go
Comment (iholder101) [reply pkg/virt-launcher/virtwrap/converter/arm64.go:27]: I see your point.

My problem with it is that it's implicit and hard to understand for someone not familiar enough with the code. For example, when looking at this file, how can I know that it aims to implement an interface? In this case it's even more important, as this struct's whole purpose (at least currently) is to implement this interface.

I think that adding it here will be valuable as it will make the code more cohesive, clearer and self-contained. But I'm willing to be convinced otherwise.

Is there a reason you think it's more valuable to not have this validation here? Are there any downsides to it?

File: pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go
Comment (iholder101) [reply pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go]: A bit similar to https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1611613115, the problem as I see it is that code can change. You are right that today `converter_test.go` might be testing it since this PR only brings refactoring and not an actual change of logic, however who knows how the code will look like tomorrow. For example, the arch converter interface can be used in other areas in the future, or the converter logic might change.

The whole concept behind unit tests is that every unit tests itself in a cohesive way. Converter has different tasks, as it converts the vmi into a libvirt XML. While it's true that it uses the arch converter, it uses it as a block box and assumes it's already working. It is not its job to test it and validate it.

One last point is that the most valuable merit of tests in general is NOT to ensure the code is working as expected (this can be validated manually as well, sometimes more easily) but to allow easy refactoring in the future and hence ease future development. Building upon an implicit assumption that the code is being tested by a different unit in an implicit way goes against this principle IMO.

File: pkg/virt-launcher/virtwrap/converter/archconverter_test.go
Comment (iholder101) [inline pkg/virt-launcher/virtwrap/converter/archconverter_test.go:24]: nit
```suggestion
var _ = Describe("Arch Converter", func() {
```

Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archconverter_test.go]: I have changed it

File: pkg/virt-launcher/virtwrap/converter/archconverter.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archconverter.go]: fixed, most likely that was vscode trying to be helpful

File: pkg/virt-launcher/virtwrap/converter/arm64.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/arm64.go:30]: No that makes sense.
The only real downside i see would be people wondering why the line is there. But i have just added a comment to it for that.

File: pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go
Comment (jschintag) [reply pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go]: I see your point, i think the difference here is what would you think of as a unit? For me in golang that has always been a package, as that is the smallest piece in golang that you can import. As such what a package should be testing is that the exported functionality is stable.
In this case that would be `converter`, as it is the exposed type. On the other hand `archConverter` exists only inside the package, it's only use is to be used by converter as a convenient tool to outsource arch-specific code. So my argument is that if `converter_test.go` does not cover it anymore, it is also no longer needed in the `converter` and should just be removed.

Comment (iholder101) [reply pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go]: That's a fair argument.
I tend to see a unit as the smaller form of an "entity" that provides some kind of a service. The reason it is valuable to look at is this way is that even inside the package, this interface can end up being used for something completely different. If we don't see the interface as a unit, then it remains untested in this case.

But in any case, I don't think it should be blocking for this PR.

**Comment by iholder101** [comment]
Thanks a lot @jschintag! Great stuff!
/lgtm
/hold cancel

**Pull request merged**
